### PR TITLE
Unit Test framework

### PR DESCRIPTION
**Pull request by yorinasub17**
Introduces a Vagrant + Serverspec based unittesting framework.

The details of the mechanism are in `tests/base.py`, under `VagrantTestCase`. The instructions for running the test cases are in the `TEST.md` file. Once you install all the requirements, tests can be run with a simple: `python setup.py test`.

### CODE REVIEW

File: TEST.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+- Testing uses Vagrant with serverspec
+- To run tests:
+    - Install vagrant
+    - Install virtualbox
+    - Install ruby and serverspec gem
```
Comment (ChristopherSchuster) [inline TEST.md]: Enable virtualization in BIOS (lenovo laptops)

File: tests/serverspecs/system_basic_spec/spec/spec_helper.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+config.close
+
+options = Net::SSH::Config.for(host, [config.path])
+
+options[:user] ||= Etc.getlogin
+
+set :host,        options[:host_name] || host
+set :ssh_options, options
+
+# Disable sudo
```
Comment (ChristopherSchuster) [inline tests/serverspecs/system_basic_spec/spec/spec_helper.rb]: Do we need these comments?

File: tests/base.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        self._setup_vagrant()
+        self._setup_cappa()
+
+    def tearDown(self):
+        self.vagrant.destroy()
+
+    def _setup_vagrant(self):
+        """Ensure a vagrant machine exists
+        """
+        self.vagrant = vagrant.Vagrant()
```
Comment (ChristopherSchuster) [inline tests/base.py]:
```
self.vagrant = vagrant.Vagrant(quiet_stdout=False)
```

Was useful for debugging purposes.

File: tests/base.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        box_status = self.vagrant.status()[0]
+        if box_status.state == vagrant.Vagrant.RUNNING:
+            self.vagrant.destroy()
+
+        self.vagrant.up()
+
+    def _setup_cappa(self):
+        """Ensure cappa is installed
```
Comment (ChristopherSchuster) [inline tests/base.py]: Nit: one line docstrings have the closing quote on the same line and end in a period. 
https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

File: TEST.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+- Testing uses Vagrant with serverspec
+- To run tests:
+    - Install vagrant
+    - Install virtualbox
```
Comment (ChristopherSchuster) [inline TEST.md]: Version 4.3.28

File: TEST.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+- Testing uses Vagrant with serverspec
+- To run tests:
+    - Install vagrant
```
Comment (ChristopherSchuster) [inline TEST.md]: Version 1.7.2

File: TEST.md (hunk 1 above)
Comment (ChristopherSchuster) [inline TEST.md]: in .ruby-version(https://gist.github.com/fnichol/1912050)
```
ruby-2.1.3
```
in Gemfile (http://bundler.io/)
```
source "https://rubygems.org"
gem 'serverspec', '= 2.17.0'

File: TEST.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+- Testing uses Vagrant with serverspec
+- To run tests:
```
Comment (ChristopherSchuster) [inline TEST.md]: Enable virtualization in BIOS (lenovo laptops)

File: tests/serverspecs/system_basic_spec/spec/spec_helper.rb (hunk 2 above)
Comment (yorinasub17) [inline tests/serverspecs/system_basic_spec/spec/spec_helper.rb]: This file was autogenerated by `serverspec-init`. I think it is best not to touch it.

File: tests/base.py (hunk 4 above)
Comment (yorinasub17) [inline tests/base.py]: Done

File: TEST.md (hunk 1 above)
Comment (yorinasub17) [inline TEST.md]: I have 2.1.2... :-/

**Comment by ChristopherSchuster** [comment]
LGTM.

**Pull request merged**